consensus/istanbul/backend/engine_test.go

```diff
@@ -1327,3 +1336,95 @@ func TestGovernance_Votes(t *testing.T) {
+	// Test that ReaderEngine (Params(), ParamsAt(), UpdateParams()) works.
+	type vote = map[string]interface{}
+	type expected = map[string]interface{} // expected (subset of) governance items
+	type testcase struct {
+		length   int // total number of blocks to simulate
+		votes    map[int]vote
+		expected map[int]expected
+	}
+
+	testcases := []testcase{
+		{
+			9,
+			map[int]vote{
+				1: {"governance.unitprice": uint64(17)},
+			},
+			map[int]expected{
+				0: {"governance.unitprice": uint64(1)},
+				1: {"governance.unitprice": uint64(1)},
+				2: {"governance.unitprice": uint64(1)},
+				3: {"governance.unitprice": uint64(1)},
+				4: {"governance.unitprice": uint64(1)},
+				5: {"governance.unitprice": uint64(1)},
+				6: {"governance.unitprice": uint64(17)},
+				7: {"governance.unitprice": uint64(17)},
+				8: {"governance.unitprice": uint64(17)},
+			},
+		},
+	}
+
+	var configItems []interface{}
+	configItems = append(configItems, proposerPolicy(params.WeightedRandom))
+	configItems = append(configItems, proposerUpdateInterval(1))
+	configItems = append(configItems, epoch(3))
+	configItems = append(configItems, governanceMode("single"))
+	configItems = append(configItems, minimumStake(new(big.Int).SetUint64(4000000)))
+	configItems = append(configItems, istanbulCompatibleBlock(new(big.Int).SetUint64(0)))
+	configItems = append(configItems, blockPeriod(0)) // set block period to 0 to prevent creating future block
+	stakes := []uint64{4000000, 4000000, 4000000, 4000000}
+
+	for _, tc := range testcases {
+		// Create test blockchain
+		chain, engine := newBlockChain(4, configItems...)
+
+		oldStakingManager := reward.GetStakingManager()
+		stakingInfo := makeFakeStakingInfo(0, nodeKeys, stakes)
+		reward.SetTestStakingManagerWithStakingInfoCache(stakingInfo)
+
+		var previousBlock, currentBlock *types.Block = nil, chain.Genesis()
+
+		// Create blocks with votes
+		for num := 0; num <= tc.length; num++ {
+			// Validate current params with Params() and CurrentSetCopy().
+			// Check that both returns the expected result.
+			assertMapSubset(t, tc.expected[num], engine.governance.Params().StrMap())
+			assertMapSubset(t, tc.expected[num], engine.governance.CurrentSetCopy())
+
+			// Place a vote if a vote is scheduled in upcoming block
+			// Note that we're building (head+1)'th block here.
+			for k, v := range tc.votes[num+1] {
+				ok := engine.governance.AddVote(k, v)
+				assert.True(t, ok)
+			}
+
+			// Create a block
+			previousBlock = currentBlock
+			currentBlock = makeBlockWithSeal(chain, engine, previousBlock)
+			_, err := chain.InsertChain(types.Blocks{currentBlock})
+			assert.NoError(t, err)
+
+			// Load parameters for the next block
+			err = engine.governance.UpdateParams()
+			assert.NoError(t, err)
+		}
+
+		// Validate historic parameters with ParamsAt() and ReadGovernance().
+		// Check that both returns the expected result.
+		for num := 0; num <= tc.length; num++ {
```

Comment:
Do we need to run the loop when num=9?

Reply:
Fixed in [97db580](https://github.com/klaytn/klaytn/pull/1380/commits/97db5809088b0653a0b346c69fcb896de60642e8).